src/server/hll_family.cc

```diff
@@ -0,0 +1,220 @@
+                "we are only handling types that are integral types in the return types from "
+                "here");
+  if (result) {
+    (*cntx)->SendLong(result.value());
+  } else {
+    switch (result.status()) {
+      case OpStatus::WRONG_TYPE:
+        (*cntx)->SendError(kWrongTypeErr);
+        break;
+      case OpStatus::OUT_OF_MEMORY:
+        (*cntx)->SendError(kOutOfMemory);
+        break;
+      case OpStatus::INVALID_VALUE:
+        (*cntx)->SendError(HllFamily::kInvalidHllErr);
+        break;
+      default:
+        (*cntx)->SendLong(0);  // in case we don't have the value we should just send 0
+        break;
+    }
+  }
+}
+
+HllBufferPtr StringToHllPtr(string* hll) {
+  return {.hll = (unsigned char*)hll->data(), .size = hll->size()};
+}
+
+void ConvertToDenseIfNeeded(string* hll) {
+  if (isValidHLL(StringToHllPtr(hll)) == HLL_VALID_SPARSE) {
+    string new_hll;
+    new_hll.resize(getDenseHllSize());
+    int result = convertSparseToDenseHll(StringToHllPtr(hll), StringToHllPtr(&new_hll));
+    DCHECK_EQ(result, 0);
+    *hll = std::move(new_hll);
+  }
+}
+
+OpResult<int> AddToHll(const OpArgs& op_args, string_view key, CmdArgList values) {
+  auto& db_slice = op_args.shard->db_slice();
+
+  string hll;
+
+  try {
+    auto [it, inserted] = db_slice.AddOrFind(op_args.db_cntx, key);
+    if (inserted) {
+      hll.resize(getDenseHllSize());
+      createDenseHll(StringToHllPtr(&hll));
+    } else if (it->second.ObjType() != OBJ_STRING) {
+      return OpStatus::WRONG_TYPE;
+    } else {
+      it->second.GetString(&hll);
+      ConvertToDenseIfNeeded(&hll);
+    }
+
+    int updated = 0;
+    for (const auto& value : values) {
+      int added = pfadd(StringToHllPtr(&hll), (unsigned char*)value.data(), value.size());
+      if (added < 0) {
+        return OpStatus::INVALID_VALUE;
+      }
+      updated += added;
+    }
+
+    db_slice.PreUpdate(op_args.db_cntx.db_index, it);
+    it->second.SetString(hll);
+    db_slice.PostUpdate(op_args.db_cntx.db_index, it, key, !inserted);
+
+    return std::min(updated, 1);
+  } catch (const std::bad_alloc&) {
+    return OpStatus::OUT_OF_MEMORY;
+  }
+}
+
+void PFAdd(CmdArgList args, ConnectionContext* cntx) {
+  string_view key = ArgS(args, 0);
+  args.remove_prefix(1);
+
+  auto cb = [&](Transaction* t, EngineShard* shard) {
+    return AddToHll(t->GetOpArgs(shard), key, args);
+  };
+
+  Transaction* trans = cntx->transaction;
+  OpResult<int> res = trans->ScheduleSingleHopT(std::move(cb));
+  HandleOpValueResult(res, cntx);
+}
+
+OpResult<int64_t> CountHllsSingle(const OpArgs& op_args, string_view key) {
+  auto& db_slice = op_args.shard->db_slice();
+
+  OpResult<PrimeIterator> it = db_slice.Find(op_args.db_cntx, key, OBJ_STRING);
+  if (it.ok()) {
+    string hll;
+    // TODO: We should provide a GetView() method in compact object to avoid some allocations:
+    // string_view GetView(string& scratch);
+    it.value()->second.GetString(&hll);
```

Comment:
We already have `GetSlice()`. We can leave a TODO: rename GetSlice to GetView 🙂

Reply:
this is the line `if ((taglen_ == 0 && encoded.size() < (1 << 15))) {`

Reply:
Sorry that I missed this line!
Do you mean that I should modify this line globally, for strings which are not HLL? I'm surprised by this, as the current number (2^15) must have been tested rigorously. Is there a way to specify a different value for custom `SetString()` calls? Or did you really mean I should change this value for everyone?

Reply:
It has not been tested rigorously. It's a stupid over-optimization that I implemented more than a year ago. It really matters only for small strings, for strings of length 1K and more it's not important. So setting 8K limit is absolutely fine.